float32 x_pos
float32 y_pos